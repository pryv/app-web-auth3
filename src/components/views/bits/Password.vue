<template>
  <div>
    <v-text-field
      :id="id"
      :value="value"
      :appendIcon="visiblePass ? 'lock_open' : 'lock'"
      :type="visiblePass ? 'text' : 'password'"
      :rules="[rules.required]"
      :label="label"
      @input="$emit('input', $event)"
      @click:append="() => (visiblePass = !visiblePass)"/>

    <v-text-field
      v-if="confirmation"
      :id="`${id}Confirmation`"
      v-model="repass"
      :type="visiblePass ? 'text' : 'password'"
      :rules="[matchPassword]"
      :label="`${label} confirmation`"/>
  </div>
</template>

<script>
export default {
  props: {
    confirmation: { type: Boolean, default: false },
    value: { type: String, default: '' },
    label: { type: String, default: 'Password' },
    id: { type: String, default: 'password' },
  },
  data: () => ({
    visiblePass: false,
    repass: '',
    rules: {
      required: value => !!value || 'Password is required.',
    },
  }),
  computed: {
    matchPassword: function () {
      return this.value === this.repass || 'Password confirmation does not match.';
    },
  },
};
</script>
