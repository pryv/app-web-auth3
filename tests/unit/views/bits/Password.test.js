import { shallowMount } from '@vue/test-utils';
import Password from '@/components/views/bits/Password';
import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.use(Vuetify);

describe('Password.test.js', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Password, {
      mocks: {
        $t: (key) => key,
      },
    });
  });

  it('renders correctly (snapshots matching)', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('asks only for password by default', () => {
    // Renders password
    const password = wrapper.find('#password');
    expect(password.exists()).toBe(true);
    // Does not render password confirmation
    const confirmation = wrapper.find('#passConfirmation');
    expect(confirmation.exists()).toBe(false);
  });

  it('asks for password and confirmation if needed', () => {
    wrapper.setProps({
      confirmation: true,
    });
    // Renders password
    const password = wrapper.find('#password');
    expect(password.exists()).toBe(true);
    // Also renders password confirmation
    const confirmation = wrapper.find('#passConfirmation');
    expect(confirmation.exists()).toBe(true);
  });

  it('validates password correctly', () => {
    const requiredRule = wrapper.vm.rules.required;
    expect(requiredRule()).toBe('Password is required');
    expect(requiredRule('notEmpty')).toBe(true);
  });

  it('validates password confirmation correctly', () => {
    wrapper.setProps({
      value: '1234',
    });
    wrapper.setData({
      repass: '1234',
    });
    // Password confirmation matches
    expect(wrapper.vm.matchPassword).toBe(true);
    wrapper.setProps({
      value: '1234',
    });
    wrapper.setData({
      repass: '4321',
    });
    // Password confirmation does not match
    expect(wrapper.vm.matchPassword).toBe('Password confirmation does not match');
  });
});
