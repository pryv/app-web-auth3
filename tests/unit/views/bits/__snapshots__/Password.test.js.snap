// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password.test.js renders correctly (snapshots matching) 1`] = `
<div>
  <v-text-field-stub
    appendicon="lock"
    id="password"
    label="Password"
    rules="function required(value) {
          return !!value || 'Password is required.';
        }"
    type="password"
    value=""
  />
   
  <!---->
</div>
`;
